beacon-chain/p2p/gossipsub_peer_crawler.go

```diff
@@ -0,0 +1,515 @@
+
+func NewGossipsubPeerCrawler(
+	service *Service,
+	dv5 ListenerRebooter,
+	crawlTimeout time.Duration,
+	crawlInterval time.Duration,
+	maxConcurrentPings int,
+	peerFilter gossipsubcrawler.PeerFilterFunc,
+	scorer PeerScoreFunc,
+) (*GossipsubPeerCrawler, error) {
+	if service == nil {
+		return nil, errors.New("service is nil")
+	}
+	if dv5 == nil {
+		return nil, errors.New("dv5 is nil")
+	}
+	if crawlTimeout <= 0 {
+		return nil, errors.New("crawl timeout must be greater than 0")
+	}
+	if crawlInterval <= 0 {
+		return nil, errors.New("crawl interval must be greater than 0")
+	}
+	if maxConcurrentPings <= 0 {
+		return nil, errors.New("max concurrent pings must be greater than 0")
+	}
+	if peerFilter == nil {
+		return nil, errors.New("peer filter is nil")
+	}
+	if scorer == nil {
+		return nil, errors.New("peer scorer is nil")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	g := &GossipsubPeerCrawler{
+		ctx:                ctx,
+		cancel:             cancel,
+		crawlInterval:      crawlInterval,
+		crawlTimeout:       crawlTimeout,
+		service:            service,
+		dv5:                dv5,
+		maxConcurrentPings: maxConcurrentPings,
+		peerFilter:         peerFilter,
+		scorer:             scorer,
+	}
+	g.pingCh = make(chan enode.Node, 4*g.maxConcurrentPings)
+	g.pingSemaphore = semaphore.NewWeighted(int64(g.maxConcurrentPings))
+	g.crawledPeers = &crawledPeers{
+		g:        g,
+		byEnode:  make(map[enode.ID]*peerNode),
+		byPeerId: make(map[peer.ID]*peerNode),
+		byTopic:  make(map[gossipsubcrawler.Topic]map[peer.ID]struct{}),
+	}
+	return g, nil
+}
+
+func (g *GossipsubPeerCrawler) PeersForTopic(topic gossipsubcrawler.Topic) []*enode.Node {
+	g.crawledPeers.mu.RLock()
+	defer g.crawledPeers.mu.RUnlock()
+
+	peerIDs, ok := g.crawledPeers.byTopic[topic]
+	if !ok {
+		return nil
+	}
+
+	var peerNodes []*peerNode
+	seen := make(map[enode.ID]bool)
+	for peerID := range peerIDs {
+		peerNode, ok := g.crawledPeers.byPeerId[peerID]
+		if !ok {
+			continue
+		}
+		if peerNode.isPinged && g.peerFilter(peerNode.node) {
+			// Skip if we've already seen this enode ID
+			if seen[peerNode.id] {
+				continue
+			}
+			seen[peerNode.id] = true
+			peerNodes = append(peerNodes, peerNode)
+		}
+	}
+
+	// Sort peerNodes in descending order of their scores.
+	sort.Slice(peerNodes, func(i, j int) bool {
+		scoreI := g.scorer(peerNodes[i].peerID)
+		scoreJ := g.scorer(peerNodes[j].peerID)
+		return scoreI > scoreJ
+	})
+
+	nodes := make([]*enode.Node, len(peerNodes))
+	for i, pn := range peerNodes {
+		nodes[i] = pn.node
+	}
+
+	return nodes
+}
+
+func (g *GossipsubPeerCrawler) RemovePeerId(peerID peer.ID) {
+	g.crawledPeers.removePeerId(peerID)
+}
+
+func (g *GossipsubPeerCrawler) RemoveTopic(topic gossipsubcrawler.Topic) {
+	g.crawledPeers.removeTopic(topic)
+}
+
+// Start runs the crawler's loops in the background.
+func (g *GossipsubPeerCrawler) Start(te gossipsubcrawler.TopicExtractor) error {
```

Comment:
Why setting the topic extractor here and not in the `NewGossipsubPeerCrawler` function?